### PR TITLE
Add url twig extension

### PR DESCRIPTION
**Pull request by luca-rath**
This PR adds some twig filters for parsing and transforming urls.
Documentation is still missing.

### CODE REVIEW

File: src/UrlTwigExtension.php [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+     * @return array
+     *
+     * @throws \InvalidArgumentException
+     */
+    private static function parseUrl($url)
+    {
+        $parsedUrl = parse_url($url);
+
+        if (false === $parsedUrl) {
+            throw new \InvalidArgumentException(sprintf('URL (%s) is malformed!', $url));
```
Comment (alexander-schranz) [inline src/UrlTwigExtension.php]: twig extension should not throw exception as you can't catch exceptions in twig return empty string instead

File: src/UrlTwigExtension.php [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+{
+    const DEFAULT_PROTOCOL = 'http';
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getFilters()
+    {
+        return [
+            new \Twig_SimpleFilter('url_sanitize', [$this, 'sanitizeUrl']),
```
Comment (alexander-schranz) [inline src/UrlTwigExtension.php]: would call it `url_format`

File: src/UrlTwigExtension.php (hunk 1 above)
Comment (luca-rath) [reply src/UrlTwigExtension.php]: doesn't twig ignore exceptions in production but show them in dev environment? with an exception developers would see that they made a mistake.

File: src/UrlTwigExtension.php [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+            new \Twig_SimpleFilter('url_protocol', [$this, 'getProtocol']),
+            new \Twig_SimpleFilter('url_host', [$this, 'getHost']),
+            new \Twig_SimpleFilter('url_port', [$this, 'getPort']),
+            new \Twig_SimpleFilter('url_domain', [$this, 'getDomain']),
```
Comment (alexander-schranz) [inline src/UrlTwigExtension.php]: would remove domain as its not clear for a developer what it should return

Comment (luca-rath) [reply src/UrlTwigExtension.php]: maybe find a better name? for me this filter would be very useful.

File: src/UrlTwigExtension.php [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+    /**
+     * Sanitizes url.
+     *
+     * @param $url string
+     *
+     * @return string|null
+     *
+     * @throws \InvalidArgumentException
+     */
+    public function sanitizeUrl($url)
```
Comment (alexander-schranz) [inline src/UrlTwigExtension.php]: should have flags then you could implement it like this:

```php
public function sanitizeUrl($url, $flags)
{
    $flags = array_merge([
        'scheme' => true,
        'user' => true,
        'pass' => true,
        'host' => true,
        'port' => true,
        'path' => true,
        'query' => true,
        'fragment' => true,
    ], $flags);

    $parseUrl = ...

    $url = '';

    foreach ($flags as $flag) {
         if (!$flag) {
               // If path is false there should be no query and no fragment
               if ($url) {
                    break;
              }

               continue;
         }

         switch ($flag) {
                case 'scheme':
                      $url .= $parseUrl[$flag] . '//';
                      break;
                case 'port':
                      $url .= 'port' . $parseUrl[$flag];
                      break;
                default:
                      $url .= $parseUrl[$flag];
                      break;
         }
    }

    return $url;
}
```

e.g. usage:

```twig
{{ url|url_format({scheme: false, path: false }) }}
```

File: src/UrlTwigExtension.php (hunk 1 above)
Comment (alexander-schranz) [reply src/UrlTwigExtension.php]: no also on production the exception is not ignored

File: README.md [hunk 5]
```diff
@@ -27,6 +27,12 @@ The [web intl twig extension](docs/intl.md) gives you a simple and efficient way
 
 [More](docs/intl.md)
 
+### 3. Url
```
Comment (alexander-schranz) [inline README.md]: add it as 5.

File: src/UrlTwigExtension.php (hunk 4 above)
Comment (luca-rath) [reply src/UrlTwigExtension.php]: Ok. That makes all the other filters unnecessary.

File: docs/url.md [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+**yml**
+
+```yml
+services:
+    app.twig.web_url:
+        class: Massive\Component\Web\UrlTwigExtension
+        tags:
+            - { name: twig.extension }
+```
```
Comment (alexander-schranz) [inline docs/url.md]: Add `## Usage` with `url_format` example

**Pull request merged**